ReSharper.FSharp/src/FSharp/FSharp.Psi.Features/src/AICore/FSharpFileSummarizer.fs

```diff
@@ -1,36 +1,271 @@
+            match child with
+            | :? IFSharpTreeNode as treeNode ->
+                try
+                    treeNode.Accept(x, context)
+                with _ -> ()
+            | _ -> ()
+
+    override x.VisitGlobalNamespaceDeclaration(namespaceDecl, context) =
+        context.AddEntity(namespaceDecl, "namespace global")
+        x.VisitNode(namespaceDecl, context)
+
+    override x.VisitNamedNamespaceDeclaration(namespaceDecl, context) =
+        // Currently, we're not adding a namespace to the scope to not duplicate its name in tokens
+        context.AddEntity(namespaceDecl, "namespace " + namespaceDecl.QualifiedName)
+        x.VisitNode(namespaceDecl, context)
+
+    override x.VisitNamedModuleDeclaration(moduleDecl, context) =
+        // Currently, we're not adding a top-level module to the scope to not duplicate its name in tokens
+        context.AddEntity(moduleDecl, "module " + moduleDecl.ClrName)
+        x.VisitNode(moduleDecl, context)
+
+    override x.VisitNestedModuleDeclaration(moduleDecl, context) =
+        let scopeRepr =
+            match context.CurrentScope with
+            | null -> "module " + moduleDecl.SourceName
+            | _ -> "." + moduleDecl.SourceName
+
+        use _ = context.OpenScope(moduleDecl, scopeRepr)
+        x.VisitNode(moduleDecl, context)
+
+    override x.VisitFSharpTypeDeclaration(typeDecl: IFSharpTypeDeclaration, context) =
+        let inheritMembers = typeDecl.TypeOrInterfaceInheritMembers
+        let inheritsRepr =
+            if Seq.isEmpty inheritMembers then "" else
+            let inherits =
+                inheritMembers
+                |> Seq.map _.TypeName
+                |> Seq.filter isNotNull
+                |> Seq.map formatTypeReferenceName
+                |> String.concat ", "
+            if inherits = "" then "" else $"inherit {inherits}"
+
+        let typeRepr = formatTypeOrExtDeclaration typeDecl
+
+        use _ = context.OpenScope(typeDecl, $"type {typeRepr}", inheritsRepr)
+        x.VisitNode(typeDecl, context)
+
+    override x.VisitExceptionDeclaration(exceptionDecl, context) =
+        use _ = context.OpenScope(exceptionDecl, $"exception {exceptionDecl.SourceName}")
+        x.VisitNode(exceptionDecl, context)
+
+    override x.VisitTypeExtensionDeclaration(extensionDecl: ITypeExtensionDeclaration, context) =
+        let extensionRepr = formatTypeOrExtDeclaration extensionDecl
+
+        use _ = context.OpenScope(extensionDecl, $"type {extensionRepr}", "with")
+        x.VisitNode(extensionDecl, context)
+
+    override x.VisitInterfaceImplementation(interfaceImpl, context) =
+        let typeName = interfaceImpl.TypeName
+        if isNull typeName then () else
+
+        let interfaceRepr = formatTypeReferenceName typeName
+
+        use _ = context.OpenScope(interfaceImpl, $"interface {interfaceRepr}")
+        x.VisitNode(interfaceImpl, context)
+
+    override x.VisitTopBinding(binding, context) =
```

Comment:
We should really try to fix the hierarchy instead of adding more repeating code like this :(